### ISSUE TITLE
Look into ReadyToRun for macOS

### ISSUE THREAD

**Issue by rolfbjarne** [xamarin/xamarin-macios#13873]
We should look into restoring support for (partial) AOT for macOS in .NET.

It looks like this is called ReadyToRun in the CoreCLR world.

Doc: https://learn.microsoft.com/en-us/dotnet/core/deploying/ready-to-run

**Comment by filipnavara** [comment]
Note that the framework assemblies are already shipped with precompiled R2R images. That's why assemblies like System.Collections.dll from the framework end up different for osx-x64 and osx-arm64 builds (mostly visible in universal app).

**Comment by rolfbjarne** [comment]
It doesn't look like we'll have time to look into this for .NET 7, so I'm postponing until .NET 8.

**Comment by rolfbjarne** [comment]
This seems relatively trivial to implement, I got a very simple hello world working like this:

https://gist.github.com/rolfbjarne/6a185790ab8fbc4fb6663aaac9d31fdd

There are two issues that prevents it from working out-of-the-box:

1. The ready to run targets get confused with assemblies whose RelativePath metadata is set to a path that includes any directory components, because the actual compilation target writes output files to the publish directory, but then there is other logic that uses RelativePath to compute the expected output path - and things break when those don't match.
2. We have to update the RelativePath of the *.r2r.dll file the ready to run compilation produces.

**Comment by rolfbjarne** [comment]
I've filed https://github.com/dotnet/runtime/issues/79902 to see if the issue with RelativePath metadata can be fixed, or if we'll have to come up with a workaround.

**Comment by rolfbjarne** [comment]
> @rolfbjarne Is there another ticket for this then? Or is this already possible?

https://github.com/xamarin/xamarin-macios/blob/main/docs/nativeaot.md
https://devblogs.microsoft.com/dotnet/dotnet-8-performance-improvements-in-dotnet-maui/#nativeaot-on-ios

**Comment by xfortin-devolutions** [comment]
Thx!

**Comment by MaxwellDAssistek** [comment]
Is it possible to do ReadyToRun on macOS now? We are unable to go for full NativeAOT and are looking into ReadyToRun as an alternative to staying interpreter only.

**Comment by rolfbjarne** [comment]
> Is it possible to do ReadyToRun on macOS now? We are unable to go for full NativeAOT and are looking into ReadyToRun as an alternative to staying interpreter only.

No, no we never implemented ReadyToRun support, because we implemented NativeAOT instead.

Not sure what you mean about "interpreter only", macOS doesn't have an interpreter, only a JIT.

**Comment by MaxwellDAssistek** [comment]
To my knowledge, due to platform restrictions, JIT is not possible for macOS/iOS apps, so the only two options are either full NativeAOT or `<UseInterpreter>true</UseInterpreter>`.

**Comment by rolfbjarne** [comment]
JIT is not possible on iOS, but it works fine on macOS (note that there's a difference between "macOS" vs "Mac Catalyst", this issue is about "macOS", not "Mac Catalyst" - you can check this in the TargetFramework value in your csproj, it'll be `net8.0-macos` for "macOS" and `net8.0-maccatalyst` for "Mac Catalyst")

The interpreter is available on iOS, but ReadyToRun is not (NativeAOT is though). Note that on iOS we also have Mono's AOT compiler, which has been around for quite a few years now.

**Comment by rolfbjarne** [comment]
Reopening, it looks like we might do this for the .NET 10 or 11 timeframe.

**Comment by rolfbjarne** [comment]
This is working for all platform in .NET 11.
